Performance Optimization: De-duplicating Derived Function Traces

rai-nhdaly opened the pull request:
We collect the dependencies of a Derived Function as a
Vector{DependencyKey}. It's important that we maintain the _order_ of
the dependencies, because we need to check their validity in order when
there are changes, since a change in an earlier dependecy may invalidate
the _keys_ of a later dependency (consider how deleting an item from a
set affects an aggregating derived function such as `sum()`).

However, as a performance optimization, we do not need to keep
_duplicate_ keys, so we also maintain a Set{DependencyKey} for
deduplicating the keys as they're recorded. Using an unordered,
hash-based Set + a Vector together provides the best of both worlds.
The set should be unordered and hash-based because we want quick
membership verification (in O(1)-time), which `Set` is, and is discarded
at the end of recording the dependencies.

This commit changes dependency tracking to maintain a Vector AND a Set
used to deduplicate the dependencies as they're recorded.

----------------

For example, this commit reduced the memory usage of our Salsa cache
from `1.9 MiB` down to `1.2 MiB`.

rai-nhdaly commented:
:) Thanks for your help, Hung! :)

mabokhamis commented:
Thanks @rai-nhdaly ! Looks great! I don't yet completely understand the concept of __order__ of key dependencies: Where does this order comes from? I have been thinking of dependecies as just a directed graph where neighbors of a vertex are unordered. However I can learn about this later without holding off this PR.

mabokhamis commented:
..unless the order is meant to be a topological order of all descendants (not just immediate neighbors) of a certain node in the dependency graph, in which case it makes perfect sense :-)

rai-nhdaly commented:
:) No it is indeed immediate neighbors. This part is pretty subtle! Todd and I discovered it together. You do indeed need to keep the functions dependencies _in order_, because the dependencies a function has are a _temporal_ phenomenon, and they can _change_ if an input changes, and so a change in an earlier dependency may **invalidate** the call to a later dependency, and you may then end up doing a whole lot of work to compute some garbage value.

Consider this situation:
```julia
# For sake of argument, imagine that there are actually thousands of students.
@input all_students: ["april", "nathan", ..., "mahmoud"]
@input student_grade: {"april"=>3.5, "nathan" => 2.5, ..., "mahmoud" => 3.0}

# Some arbitrary transformation of the inputs; take only the second half of alphabet
@derived all_students_n_through_z(state) = [uppercase(n) for n in all_students]

@derived function mean_grade_n_through_z(state)
    mean(state.student_grade[s] for s in all_students_n_through_z(state))
end
```
If you call this, you'll get a dependency tree like this:
```julia
mean_grade_n_through_z()
├── all_students_n_through_z()
├── student_grade["nathan"]
├─ ...
└── student_grade["mahmoud"]
```
```julia
all_students_n_through_z()
└── all_students
```

Now, let's say you _delete `"nathan"` from both inputs_, leading to this new input state:
```julia
@input all_students: ["alice",..., "mahmoud"]
@input student_grade: {"alice",..., "mahmoud" => 3.0}
```

If you then call `mean_grade_n_through_z(state)` again, you need to determine whether or not `mean_grade_n_through_z(state)` _actually_ needs to be recomputed. To answer that question, it's not enough to just check whether any of its inputs were modified, we also want to check whether the modification actually _changes the value_. If none of the dependencies actually produce different values, we do not need to run `mean_grade_n_through_z()` again.

For example, if we had deleted only `"april"`, we know we need to rerun `all_students_n_through_z()`, since its input (`all_students`) has changed. HOWEVER, `all_students_n_through_z()` ends up returning the same exact value, since `"april"` was already filtered out. (We also check all of the other dependencies but they have not been modified (we track changes per-key for maps).) So we _do not_ need to re-run `mean_grade_n_through_z(state)`, because none of its inputs have changed.  This is called the Early Exit Optimization. 

Coming back to the example of deleting `"nathan"`: Here, we know that it _will_ affect the result of `all_students_n_through_z()` meaning we _will_ need to re-run the program. We will do so immediately, and cease checking all our dependencies. If we did it in the other order (arbitrarily), and checked the _other_ dependencies first, we would get a `KeyError` when we attempt to see whether the `student_grade["nathan"]` dependency has been modified, because we've deleted it (along with its previous value information). Also, we could end up checking all those thousands of inputs twice: once to confirm if their value has changed, and then _again_ when we actually do rerun the function and need to get their value.

But when we go in the original order, we'll first see that `all_students_n_through_z()` has changed, so we immediately know we need to rerun the function.

-----------

So, that's why. It falls out of us checking for input changes _lazily_, only when the user actually requests the output, on demand.
And, at least, this is the way we're doing things right now. Of course a lot of this is open to changing! For example, we could probably just _check_ whether the key we're looking for is still there in the Input, and if not, consider that an input change. That might actually be enough to let us check the dependencies in arbitrary order... Though I _think_ there is still a bit of a performance benefit to doing it this way?

The more complicated case would be if the secondary input had _also_ been a derived value, and end up going through a bunch of expensive work to compute it after the input change, only to find that the input change also changed the function so that it no longer _wants_ the value we just spent all that work computing!

Actually, yeah, that's a better example: Imagine you have this situation:
```julia
B = 10
A <- if B > 0   C else D
C <- B * 2 + 3 * 4 - 5    # Some expensive computation
```
We change `B = -10`.

We know this means we need to recompute A. But imagine we check `A`'s dependencies in an arbitrary order, and we happen to check `C` first. To actually see whether C has changed or not, based on the change to B, we need to _run the expensive computation._ So then we do all that, and see that indeed `C` _has_ changed, so now we know we need to rerun `A`. BUT when we do that, we find that actually, OOPS, we **don't even _need_ `C` anymore**, so that was wasted work!

That's a better reason. :)

--------

Hopefully all that makes sense? Sorry for the stream of consciousness; i feel like both those examples are useful (and might make good documentation?) so i left them in my comment :)

mabokhamis commented:
Awesome! The examples are very helpful: So the order of dependencies `x1, x2, ..., x_n` of some node `y` is the order in which the dependencies `x1, x2,...` are used during the computation of `y`, which is a very natural thing to do, and minimizes the amount of work needed to check for changes, as shown by the examples.

I do feel guilty though for making you type all of that in the middle of the night :-) Would probably feel less guilty if the above ended up being part of the documentation +1
